### ISSUE TITLE
align module instance named port connections

### ISSUE THREAD

**Issue by fangism** [google/verible#313]
Like this:

```systemverilog
module m;
  foo #(
    ...
  ) bar (
    .a   (a),
    .aa  (aa),
    .aaa (aaa)
  );
endmodule
```

If a port is abbreviated with just `.aa`, then there is no need for that line to align with anything else.

Recommend following the example:

https://github.com/google/verible/blob/85f8ff8134957d220400e29b05036318523ea22e/verilog/formatting/align.cc#L80

I expect the handling around new column markers in the case of named port connections to be simpler.

**Comment by fangism** [comment]
b/152805837

**Comment by Ciprian167** [comment]
I would like to give this a try. @fangism, could you please assign it to me?

**Comment by fangism** [comment]
Assigned, thanks for volunteering.

**Comment by fangism** [comment]
Quickly checking in, have you been able to make some progress or have any questions?

**Comment by Ciprian167** [comment]
I didn't got the chance to start it yet. I will come back with questions soon.

**Comment by fangism** [comment]
No problem.  Make sure you sync up through at least https://github.com/google/verible/commit/bdbc19b016898410e8eb3c0b327d822f1acbd994 because I've made some interface improvements that should help with controlling alignment.

**Comment by fangism** [comment]
Just checking in again.  Have any questions?

**Comment by Ciprian167** [comment]
> Just checking in again.  Have any questions?
Sorry I didn't had much time this past month. I'm giving it try today.

**Comment by Ciprian167** [comment]
@fangism I've got a couple of questions:
1. Right now I implemented the named port scanner inside /verible/verilog/formatting/align.cc. Should I move it out in a separate file?

2. I'm trying to workout the following corner cases when interlacing wildcard connections, implicit connections and named connections. 
First, the following:
 ```systemverilog
module m;
    foo bar(.a(a), .aa, .bbb(b));
endmodule
```
Gets formatted:
```systemverilog
module m;
  foo bar (
      .a  (a),
      .aa,
      .bbb(b)
  );
endmodule
```
I guess `.a(a)` should not be aligned with `.bbb(b)`, since they are divided by an implicit connection, which is not aligned with any of them. I assume, I have to do something similar to `GetSubpartitionsBetweenBlankLines` and break partitions separated by blank lines, `.*`, and `.a`, right?

Second, the following:
 ```systemverilog
module m;
    foo bar(.a(a), .*, .bbb(b));
endmodule
```
Ouputs:
```systemverilog
module m;
  foo bar (
      .a  (a), .*,
      .bbb(b)
  );
endmodule
```
This feels like something doesn't work with `kFitOnLineElseExpand` policy. Even when not applying my alignment I get `.*` on the same line with .a(a).

**Comment by fangism** [comment]
1. It's ok to keep Verilog-specific alignment code in verilog/formatting/align.cc, it was expected to grow.

2. (various)

For now, you may treat the `.aa` implicit named connection like a common 'row': it does not participate in alignment, but rows above and below it still align with each other.  Currently, a blank line is what divides alignment groups in module port declarations, and that's a reasonable start.  What you show actually looks acceptable, with `(a)` lining up with `(b)`.

As for `.*`, there is probably logic missing for making that occupy its own partition, which explains what you're seeing.

I examine `verible-verilog-format --show_token_partition_tree` on your example and see:

```
Full token partition tree:
{ ([<auto>], policy: always-expand) @{}
  { ([<auto>], policy: fit-else-expand) @{0}, (origin: "module m;
 ...;
endmodule")
    { ([module m ;], policy: fit-else-expand, (origin: "module m;")) }
    { (>>[<auto>], policy: fit-else-expand) @{0,1}, (origin: "foo bar(.a(..., .bbb(b));")
      { (>>[foo bar (], policy: fit-else-expand, (origin: "foo")) }
      { (>>>>>>[<auto>], policy: always-expand) @{0,1,1}, (origin: ".a(a), .*, .bbb(b)")
        { (>>>>>>[. a ( a ) , .* ,], policy: fit-else-expand, (origin: ".a(a)")) }     // OOPS: .* should have started its own partition
        { (>>>>>>[. bbb ( b )], policy: fit-else-expand, (origin: ".bbb(b)")) }
      }
      { (>>[) ;], policy: uninitialized) }
    }
    { ([endmodule], policy: always-expand, (origin: "endmodule")) }
  }
}
```

I would treat `.*` handling as a separate issue to be fixed in verilog/formatting/tree_unwrapper.cc.

For now, it is acceptable in this PR to leave `.*` not handled properly.  We can fix one thing at a time.

Thanks for continuing this!

**Comment by fangism** [comment]
Separated the `.*` partitioning issue as #363

**Comment by Ciprian167** [comment]
There is a similar problem when using  comments:
```systemverilog
module m;
    foo bar(
    .a(a),
    //.aa(aa), 
    .aaa(aaa));
endmodule
```
Outputs:
```systemverilog
module m;
  foo bar (
      .a  (a),
      //.aa(aa), .aaa(aaa)
  );
endmodule
```
Which can be problematic.

**Comment by fangism** [comment]
That needs to be fixed by adding a case here:

https://github.com/google/verible/blob/1886f72085356d3169d35a8b496f97fe1c662959/verilog/formatting/tree_unwrapper.cc#L513

Let me try to fix this quickly today.

(Incidentally, the formatter runs a lexical equivalence check at the end and would bail out to prevent this sort of data corruption.)

**Comment by Ciprian167** [comment]
I suppose we need a case for `NodeEnum::kPortActualList`. I can add it and commit the changes for this issue as well.

**Comment by fangism** [comment]
Exactly.  I'll split this off as a separate issue, and assign it to you.  (I like keeping issues separate because it make it easier for developers to understand small directed fixes.)

**Comment by fangism** [comment]
Hmm, I tried to run `verible-verilog-format --format_module_instantiations` on your example with comments, and am unable to reproduce the problematic output.  There may be some interaction with your ongoing work.  I can't file a separate issue without a standalone test case, but I'll let you continue to work on it.

**Comment by Ciprian167** [comment]
You're right. I can see this only after changing this policy to `kTabularAlignment`
https://github.com/google/verible/blob/1886f72085356d3169d35a8b496f97fe1c662959/verilog/formatting/tree_unwrapper.cc#L919
